### PR TITLE
fix: atomic cas

### LINKED ISSUE #292: Race condition when JoinHandle is dropped concurrently
**Version**

latest commit on branch master (https://github.com/bytedance/monoio/commit/e7a6dc13d67831553bd55901f58b5f1fb436d625)

**Platform**

Linux 6.6

**Description**

There is a non-atomic compare-and-swap operation in [monoio::task::state::State::drop_join_handle_fast](https://github.com/bytedance/monoio/blob/e7a6dc13d67831553bd55901f58b5f1fb436d625/monoio/src/task/state.rs#L159), which can lead to race condition and use-after-free when JoinHandle is dropped concurrently.

[Proof-of-concept](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=2430f67551569920cf93f79f9710e14f) that can cause segmentation fault.

This issue originates from my attempt to integrate monoio into OpenDAL (https://github.com/apache/opendal/pull/4944), where a ci fails at random.

### PR DESCRIPTION
**Pull request by NKID00**
Fixes #292.

### CODE REVIEW

File: monoio/src/task/state.rs [hunk 1]
```diff
@@ -80,6 +80,7 @@ impl State {
         Snapshot(self.0.load(Acquire))
     }
 
+    #[allow(dead_code)]
```
Comment (Xuanwo) [inline monoio/src/task/state.rs]: Better to remove if it's dead.

**Review by Xuanwo — COMMENTED** [review]
I'm not sure if this fix is correct. Could you create a unit test for it? It should fail on the main branch but work with this PR.

**Review by ihciah — DISMISSED** [review]
Thank you! I think it is a good fix.
This problem was brought in previously when I changed the thread local style code to across-thread style to support "sync" feature. This should be nearly the same as tokio's.

File: monoio/src/task/state.rs (hunk 1 above)
Comment (ihciah) [reply monoio/src/task/state.rs]: Agreed. If there's no any usage, it should be removed. However, if it is used under some condition(some `#cfg`), this is acceptable for me.

Comment (NKID00) [reply monoio/src/task/state.rs]: The dead code is now removed.

**Review by ihciah — APPROVED** [review]

**Pull request merged**